packages/matter.js/src/protocol/interaction/SubscriptionHandler.ts

```diff
@@ -821,64 +810,56 @@ export class SubscriptionHandler {
-                                    };
-                                }),
+            if (attributes.length === 0 && events.length === 0) {
+                await messenger.sendDataReport(
+                    {
+                        suppressResponse: true, // suppressResponse true for empty DataReports
+                        subscriptionId: this.subscriptionId,
+                        interactionModelRevision: INTERACTION_MODEL_REVISION,
+                    },
+                    this.isFabricFiltered,
+                );
+            } else {
+                await messenger.sendDataReport(
+                    {
+                        suppressResponse: false, // Non empty data reports always need to send response
+                        subscriptionId: this.subscriptionId,
+                        interactionModelRevision: INTERACTION_MODEL_REVISION,
+                        attributeReportsPayload: attributes.map(({ path, schema, value, version, attribute }) => ({
+                            hasFabricSensitiveData: attribute.hasFabricSensitiveData,
+                            attributeData: {
+                                path,
+                                dataVersion: version,
+                                schema,
+                                payload: value,
                             },
-                            this.isFabricFiltered,
-                        );
-                    }
-                },
-                StatusResponseError,
-                async error => {
-                    if (error.code === StatusCode.InvalidSubscription || error.code === StatusCode.Failure) {
-                        logger.info(`Subscription ${this.subscriptionId} cancelled by peer.`);
-                        await this.cancel(false, true);
-                    } else {
-                        await this.cancel(false);
-                        throw error;
-                    }
-                },
-            );
+                        })),
+                        eventReportsPayload: events.map(({ path, schema, event, data }) => {
+                            const { eventNumber, priority, epochTimestamp, data: payload } = data;
+                            return {
+                                hasFabricSensitiveData: event.hasFabricSensitiveData,
+                                eventData: {
+                                    path,
+                                    eventNumber,
+                                    priority,
+                                    epochTimestamp,
+                                    schema,
+                                    payload,
+                                },
+                            };
+                        }),
+                    },
+                    this.isFabricFiltered,
+                );
+            }
+        } catch (e) {
+            if (StatusResponseError.is(e, StatusCode.InvalidSubscription, StatusCode.Failure)) {
```

Comment:
StatusResponse.accept(e) missing

Reply:
I think this one is OK.  `StatusResponseError.is()` tests the error class and the status code.  `accept()` would throw immediately.  We do rethrow below but we do the cancel first.

Reply:
right